File: app/src/main/java/com/security/passwordmanager/ui/passwords/PasswordListFragment.java
package com.security.passwordmanager.ui.passwords;

import android.annotation.SuppressLint;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.text.InputType;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.DimenRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.motion.widget.MotionLayout;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.security.passwordmanager.BottomSheet;
import com.security.passwordmanager.Data;
import com.security.passwordmanager.DataLab;
import com.security.passwordmanager.PasswordActivity;
import com.security.passwordmanager.R;
import com.security.passwordmanager.Support;

import java.util.List;

public class PasswordListFragment extends Fragment {

    private RecyclerView mRecyclerView;
    private PasswordAdapter mAdapter;

    private Support mSupport;
    private DataLab mDataLab;
    private List<Data> dataList;

    private BottomSheet mBottomSheet;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {

        View root = inflater.inflate(R.layout.fragment_password_list, container, false);
        mRecyclerView = root.findViewById(R.id.main_recycler_view);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        updateUI();

        mBottomSheet = new BottomSheet(requireContext(), root);

        return root;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        mSupport = Support.get(getActivity());
        mDataLab = DataLab.get(getActivity());
    }

    @Override
    public void onResume() {
        super.onResume();
        updateUI();
    }

    private void updateUI() {
        dataList = mDataLab.getDataList();

        if (mAdapter == null) {
            mAdapter = new PasswordAdapter();
            mRecyclerView.setAdapter(mAdapter);
        }
        else mAdapter.notifyDataSetChanged();
    }



    private class PasswordAdapter extends RecyclerView.Adapter<PasswordHolder> {

        private static final int OPENED_VIEW = 0xAAA;
        private int openedView = -1;

        @Override
        public int getItemViewType(int position) {
            if (position == openedView)
                return OPENED_VIEW;
            return super.getItemViewType(position);
        }

        @NonNull
        @Override
        public PasswordHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            LayoutInflater inflater = LayoutInflater.from(getActivity());
            View view = inflater
                    .inflate(R.layout.list_item_main_text_view, parent, false);

            RecyclerView infoRecyclerView = view.findViewById(R.id.list_item_recycler_view_more_info);

            if (viewType == OPENED_VIEW)
                infoRecyclerView.setVisibility(View.VISIBLE);
            else
                infoRecyclerView.setVisibility(View.GONE);

            return new PasswordHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull PasswordHolder holder, int position) {
            Data data = dataList.get(position);
            holder.bindPassword(mDataLab.getAccountList(data.getAddress()), position == openedView);

            holder.setOnClickListener(v -> {
                if (openedView == position)
                    openedView = -1;
                else {
                    int oldOpen = openedView;
                    openedView = position;
                    notifyItemChanged(oldOpen);
                }
                notifyItemChanged(position);
            });
        }

        @Override
        public int getItemCount() {
            return dataList.size();
        }
    }


    private class PasswordHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private final RecyclerView recyclerView;
        private final LinearLayout mainView;
        private final ImageView imageView;
        private final TextView textView_name, textView_url;
        private final Button button_more;
        private final MotionLayout motionLayout;

        private List<Data> mAccountList;

        public PasswordHolder(@NonNull View itemView) {
            super(itemView);
            mainView = itemView.findViewById(R.id.list_item_main_view);
            imageView = itemView.findViewById(R.id.list_item_image_view);
            textView_name = itemView.findViewById(R.id.list_item_text_view_name);
            textView_url = itemView.findViewById(R.id.list_item_text_view_url);
            button_more = itemView.findViewById(R.id.list_item_button_more);

            recyclerView = itemView.findViewById(R.id.list_item_recycler_view_more_info);
            recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));

            button_more.setOnClickListener(this);

            motionLayout = itemView.findViewById(R.id.motion_container);
        }

        public void bindPassword(List<Data> accountList, boolean isShown) {
            this.mAccountList = accountList;

            recyclerView.setAdapter(new MoreInfoAdapter(mAccountList));

            updateArrow(isShown);
            textView_name.setText(accountList.get(0).getNameWebsite());
            textView_url.setText(accountList.get(0).getAddress());

            imageView.setImageTintList(ColorStateList.valueOf(mSupport.getHeaderColor()));
            textView_name.setBackgroundColor(mSupport.getBackgroundColor());
            textView_url.setBackgroundColor(mSupport.getBackgroundColor());
            button_more.setBackgroundTintList(ColorStateList.valueOf(mSupport.getFontColor()));

            textView_name.setTextColor(mSupport.getFontColor());
        }

        @SuppressLint("NonConstantResourceId")
        @Override
        public void onClick(View v) {

            mBottomSheet.setHeading(
                    mAccountList.get(0).getNameWebsite(),
                    mAccountList.get(0).getAddress()
            );

            mBottomSheet.updateImageAndText(
                    new int[]{ R.string.edit_password, R.string.delete_password },
                    new int[]{ R.drawable.ic_outline_edit_24, R.drawable.ic_baseline_delete_24 }
            );

            mBottomSheet.setOnClickListener(BottomSheet.VIEW_EDIT, v1 -> {
                startActivity(PasswordActivity
                        .newIntent(requireContext(), mAccountList.get(0).getAddress()));
                mBottomSheet.stop();
            });

            mBottomSheet.setOnClickListener(BottomSheet.VIEW_DELETE, v1 -> {
                mDataLab.deleteData(mAccountList.get(0).getAddress());
                mBottomSheet.stop();
                updateUI();
            });

            mBottomSheet.start();
        }

        public void setOnClickListener(View.OnClickListener listener) {
            mainView.setOnClickListener(listener);
        }


        //true - стрелка вниз, false - вправо
        private void updateArrow(boolean isShown) {
            if (isShown)
                motionLayout.transitionToEnd();
            else
                motionLayout.transitionToStart();
        }
    }



    private class MoreInfoAdapter extends RecyclerView.Adapter<MoreInfoHolder> {

        private final List<Data> accountList;

        public MoreInfoAdapter(List<Data> accountList) {
            this.accountList = accountList;
        }

        @NonNull
        @Override
        public MoreInfoHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(getActivity())
                    .inflate(R.layout.list_item_more_info, parent, false);
            return new MoreInfoHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull MoreInfoHolder holder, int position) {
            Data data = accountList.get(position);
            holder.bindInfo(data);
        }

        @Override
        public int getItemCount() {
            return accountList.size();
        }
    }



    private class MoreInfoHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private final LinearLayout moreInfo;
        private final TextView accountName;
        private final ConstraintLayout login, password, comment;
        private final TextView head_login, head_password, head_comment;
        private final Button button_open_url;
        private Data data;

        private boolean isPasswordVisible;

        public MoreInfoHolder(@NonNull View itemView) {
            super(itemView);
            moreInfo = itemView.findViewById(R.id.list_item_linear_layout_more);

            accountName = itemView.findViewById(R.id.list_item_text_view_name_of_account);
            login = itemView.findViewById(R.id.list_item_text_view_login);
            password = itemView.findViewById(R.id.list_item_text_view_password);
            comment = itemView.findViewById(R.id.list_item_text_view_comment);
            button_open_url = itemView.findViewById(R.id.list_item_button_open_url);

            button_open_url.setOnClickListener(this);

            head_login = itemView.findViewById(R.id.list_item_text_view_login_head);
            head_password = itemView.findViewById(R.id.list_item_text_view_password_head);
            head_comment = itemView.findViewById(R.id.list_item_text_view_comment_head);

            button_open_url.setTextColor(Color.WHITE);

            isPasswordVisible = false;
        }

        public void bindInfo(Data data) {
            this.data = data;

            if (data.getComment().length() == 0) {
                comment.setVisibility(View.GONE);
                head_comment.setVisibility(View.GONE);
            }

            accountName.setText(data.getNameAccount());
            setTextToLayout(login, data.getLogin(), null);
            setTextToLayout(password, data.getPassword(), 129);
            setTextToLayout(comment, data.getComment(), null);

            setOnClickListener(login, data.getLogin(), false);
            setOnClickListener(password, data.getPassword(), true);
            setOnClickListener(comment, data.getComment(), false);

            moreInfo.setBackgroundColor(mSupport.getLayoutBackgroundColor());
            accountName.setTextColor(mSupport.getDarkerGrayColor());
            accountName.setBackgroundColor(mSupport.getLayoutBackgroundColor());
            setColor(login);
            setColor(password);
            setColor(comment);
            button_open_url.setBackgroundResource(mSupport.getButtonRes());

            head_login.setTextColor(mSupport.getFontColor());
            head_password.setTextColor(mSupport.getFontColor());
            head_comment.setTextColor(mSupport.getFontColor());
        }

        @SuppressLint("NonConstantResourceId")
        @Override
        public void onClick(View v) {
            if (v.getId() == R.id.list_item_button_open_url) {
                String address;
                if (data.getAddress().contains("www."))
                    address = "https://" + data.getAddress();
                else if (data.getAddress().contains("https://www.") || data.getAddress().contains("http://www."))
                    address = data.getAddress();
                else
                    address = "https://www." + data.getAddress();

                Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(address));
                startActivity(intent);
            }
        }

        private void setTextToLayout(ConstraintLayout layout, String text, Integer inputType) {
            TextView textView = layout.findViewById(R.id.field_item_text_view);
            textView.setText(text);
            if (inputType != null)
                textView.setInputType(inputType);
        }

        private void setOnClickListener(ConstraintLayout layout, String text, boolean needVisibility) {
            ImageButton copy = layout.findViewById(R.id.field_item_button_copy);
            ImageButton visibility = layout.findViewById(R.id.field_item_button_visibility);

            copy.setOnClickListener(v -> {
                ClipboardManager clipboard = (ClipboardManager)
                        requireActivity().getSystemService(Context.CLIPBOARD_SERVICE);
                ClipData clip = ClipData.newPlainText("", text);
                clipboard.setPrimaryClip(clip);
                Toast.makeText(getActivity(), "Скопированно!", Toast.LENGTH_SHORT).show();
            });

            if (needVisibility) {
                visibility.setVisibility(View.VISIBLE);
                visibility.setOnClickListener(v -> {
                    TextView textView = layout.findViewById(R.id.field_item_text_view);

                    isPasswordVisible = !isPasswordVisible;

                    if (isPasswordVisible) {
                        textView.setInputType(InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD);
                        visibility.setImageResource(R.drawable.ic_outline_visibility_off_24);
                    }
                    else {
                        textView.setInputType(129);
                        visibility.setImageResource(R.drawable.ic_visibility_24);
                    }
                });
            }
            else visibility.setVisibility(View.GONE);
        }

        private void setColor(ConstraintLayout layout) {
            TextView textView = layout.findViewById(R.id.field_item_text_view);
            ImageButton copy = layout.findViewById(R.id.field_item_button_copy);
            ImageButton visibility = layout.findViewById(R.id.field_item_button_visibility);

            textView.setTextColor(mSupport.getFontColor());
            copy.setImageTintList(ColorStateList.valueOf(mSupport.getFontColor()));
            visibility.setImageTintList(ColorStateList.valueOf(mSupport.getFontColor()));

            textView.setBackgroundColor(mSupport.getLayoutBackgroundColor());
            copy.setBackgroundColor(mSupport.getLayoutBackgroundColor());
            visibility.setBackgroundColor(mSupport.getLayoutBackgroundColor());
        }
    }
}

File: app/src/main/java/com/security/passwordmanager/view/PasswordListFragment.kt
package com.security.passwordmanager.view

import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.view.*
import android.widget.EditText
import android.widget.ImageView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import com.security.passwordmanager.*
import com.security.passwordmanager.databinding.FragmentPasswordListBinding
import com.security.passwordmanager.model.Data
import com.security.passwordmanager.view.adapters.PasswordListAdapter
import com.security.passwordmanager.viewmodel.DataViewModel

class PasswordListFragment : Fragment() {

    companion object {
        private const val OPENED_VIEW_KEY = "opened_view"
//        private const val VISIBILITIES_KEY = "visibilities"
    }

    private lateinit var binding: FragmentPasswordListBinding

    private lateinit var adapter: PasswordListAdapter
    private var recyclerCallback: RecyclerCallback? = null

    private var openedView = -1
    set(value) {
        field = value
        adapter.openedView = value
    }

    private lateinit var dataViewModel: DataViewModel
    private val dataList: ArrayList<Data> = ArrayList()


    override fun onAttach(context: Context) {
        super.onAttach(context)
        recyclerCallback = activity as RecyclerCallback
    }

    override fun onDetach() {
        super.onDetach()
        recyclerCallback = null
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        activity?.addMenuProvider (object: MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menuInflater.inflate(R.menu.main, menu)

                val searchView = menu.findItem(R.id.menu_item_search).actionView as SearchView
                drawSearchView(searchView)

                searchView.doOnQueryTextChange {
                    val list = emptyList<Data>() //dataViewModel.searchData(it)
                    openedView = -1
                    adapter.dataList = list
                }
            }

            override fun onMenuItemSelected(menuItem: MenuItem) = when (menuItem.itemId) {
                R.id.menu_item_search -> true
                else -> false
            }
        })
    }


    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {

        binding = FragmentPasswordListBinding.inflate(inflater, container, false)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        activity?.let {
            dataViewModel = DataViewModel.getInstance(this)
        }

        val openedView = savedInstanceState.getInt(OPENED_VIEW_KEY, -1)

        adapter = PasswordListAdapter(activity as AppCompatActivity)
        this.openedView = openedView

        binding.mainRecyclerView.adapter = adapter

        binding.mainRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                recyclerCallback?.onScroll(
                    when {
                        dy > 1 -> RecyclerDirection.DOWN
                        dy < 1 -> RecyclerDirection.UP
                        else -> RecyclerDirection.STOPPED
                    },
                    RecyclerState.getState(recyclerView.scrollState)
                )

                super.onScrolled(recyclerView, dx, dy)
            }

            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                recyclerCallback?.onStateChanged(RecyclerState.getState(newState))

                super.onScrollStateChanged(recyclerView, newState)
            }
        })
    }


    override fun onResume() {
        super.onResume()
//        dataList.updateAll(dataViewModel.getDataList())
        adapter.dataList = dataList
    }


    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(OPENED_VIEW_KEY, openedView)
    }

    private fun drawSearchView(searchView: SearchView) {
        val searchText = searchView.findViewById<EditText>(androidx.appcompat.R.id.search_src_text)
        val searchImage = searchView.findViewById<ImageView>(androidx.appcompat.R.id.search_button)
        val clearView = searchView.findViewById<ImageView>(R.id.search_close_btn)

        searchText.setTextColor(Color.WHITE)
        searchText.backgroundTintList = ColorStateList(Color.WHITE)
        searchImage.imageTintList = ColorStateList(Color.WHITE)
        clearView.imageTintList = ColorStateList(Color.WHITE)
    }


//    private fun String.createUrlString() = when {
//        "www." in this -> "https://$this"
//        "https://www." in this || "http://www." in this -> this
//        else -> "https://www.$this"
//    }



//    private inner class PasswordAdapter : RecyclerView.Adapter<PasswordHolder>() {
//
//        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PasswordHolder {
//            val itemBinding = MainTextViewBinding
//                .inflate(LayoutInflater.from(parent.context), parent, false)
//
//            val holder = PasswordHolder(itemBinding)
//
//            holder.setOnClickListener {
//                if (holder.adapterPosition == openedView)
//                    openedView = -1
//                else {
//                    val oldOpen = openedView
//                    openedView = holder.adapterPosition
//                    notifyItemChanged(oldOpen)
//                }
//                notifyItemChanged(holder.adapterPosition)
//            }
//
//            return holder
//        }
//
//        override fun onBindViewHolder(holder: PasswordHolder, position: Int) {
//            val data = dataList[position]
//            holder.bindPassword(data, position == openedView)
//        }
//
//        override fun getItemCount() = dataList.size
//    }
//
//
//
//
//    private inner class PasswordHolder(private val itemBinding: MainTextViewBinding):
//        RecyclerView.ViewHolder(itemBinding.root), View.OnClickListener {
//
//        private lateinit var accountList: ArrayList<Data>
//        private var moreInfoAdapter: MoreInfoAdapter? = null
//
//        private lateinit var data: Data
//
//        init {
//            itemBinding.showBottomFragment.setOnClickListener(this)
//        }
//
//
//        private fun LayoutMainHeaderBinding.getTextFields() = Pair(heading, subtitle)
//
//        private fun LayoutMainHeaderBinding.setText(headingText: String, subtitleText: String) {
//            val fields = getTextFields()
//            fields[0].text = headingText
//            fields[1].text = subtitleText
//            fields.setAll { it.setBackgroundColor(settings.backgroundColor) }
//            fields[0].setTextColor(settings.fontColor)
//        }
//
//        private fun LayoutMainHeaderBinding.getText() = Pair {
//            getTextFields()[it].text.toString()
//        }
//
//
//        fun bindPassword(data: Data, isShown: Boolean) {
//            this.data = data
//            accountList = dataViewModel.getAccountList(data) as ArrayList<Data>
//
//            moreInfoAdapter = MoreInfoAdapter(accountList)
//
//            itemBinding.run {
//                recyclerViewDescription.adapter = moreInfoAdapter
//
//                updateArrow(isShown)
//                setVisible(isShown)
//
//                when (data) {
//                    is Website -> {
//                        textView.setText(data.nameWebsite, data.address)
//
////                        Picasso.get()
////                            .load("https://yandex.ru")
////                            .error(R.drawable.website_image)
////                            .into(websiteImage)
//                    }
//                    is BankCard -> {
//                        textView.setText(data.bankName, data.cardNumber)
//                        websiteImage.imageTintList = ColorStateList.valueOf(settings.headerColor)
//                    }
//                }
//                imageView.imageTintList = ColorStateList.valueOf(settings.headerColor)
//                showBottomFragment.imageTintList = ColorStateList.valueOf(settings.fontColor)
//                showBottomFragment.setBackgroundColor(settings.backgroundColor)
//            }
//        }
//
//        private fun setVisible(visible: Boolean) {
//            val visibility = if (visible) View.VISIBLE else View.GONE
//            itemBinding.recyclerViewDescription.visibility = visibility
//        }
//
//        override fun onClick(v: View?) {
//            /** v = btnMore */
//            val text = itemBinding.textView.getText()
//
//            val bottomDialogFragment = ActionBottom.newInstance(activity as AppCompatActivity)
//            bottomDialogFragment.setHeading(text[0], text[1])
//
//            bottomDialogFragment.addView(R.drawable.edit, R.string.edit) {
//                when(data) {
//                    is Website -> startActivity(
//                        PasswordActivity.getIntent(context, data.key))
//                    is BankCard -> startActivity(
//                        BankCardActivity.getIntent(context, data.key))
//                }
//                bottomDialogFragment.dismiss()
//            }
//
//            bottomDialogFragment.addView(R.drawable.copy, R.string.copy_info) {
//                dataViewModel.copyAccountList(accountList)
//                bottomDialogFragment.dismiss()
//            }
//
//            bottomDialogFragment.addView(R.drawable.delete, R.string.delete_password) {
//                dataViewModel.deleteRecords(data)
//
//                bottomDialogFragment.dismiss()
//                adapter?.notifyItemRemoved(adapterPosition)
//            }
//            openBottomSheet(bottomDialogFragment)
//        }
//
//        fun setOnClickListener(l: View.OnClickListener) =
//            itemBinding.mainLayout.setOnClickListener(l)
//
//
//        //true - стрелка вниз, false - вправо
//        private fun updateArrow(isShown: Boolean) = if (isShown) {
//            itemBinding.motionContainer.transitionToEnd()
//        }
//        else {
//            itemBinding.motionContainer.transitionToStart()
//        }




//        private inner class MoreInfoAdapter(private val accountList: ArrayList<Data>):
//            RecyclerView.Adapter<MoreInfoHolder>() {
//
//            override fun getItemViewType(position: Int) =
//                accountList[position].type.number
//
//            override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MoreInfoHolder {
//                val layoutInflater = LayoutInflater.from(context)
//
//                val moreInfoBinding = when(viewType) {
//                    DataType.BANK_CARD.number -> MoreBankCardBinding
//                        .inflate(layoutInflater, parent, false)
//
//                    else -> MoreWebsiteBinding
//                        .inflate(layoutInflater, parent, false)
//                }
//
//                return MoreInfoHolder(moreInfoBinding)
//            }
//
//            override fun onBindViewHolder(holder: MoreInfoHolder, position: Int) {
//                val data = accountList[position]
//                holder.bindInfo(data, position)
//            }
//
//            override fun getItemCount() =
//                accountList.size
//        }
//
//
//
//        private inner class MoreInfoHolder(private val moreInfoBinding: ViewBinding):
//            RecyclerView.ViewHolder(moreInfoBinding.root) {
//
//            private lateinit var data: Data
//            private var pos = 0
//
//            private val bottomDialogFragment = ActionBottom.newInstance(activity as AppCompatActivity)
//
//            private var isPasswordVisible = false
//
//            fun TextView.setTextColor() =
//                setTextColor(settings.fontColor)
//
//            private fun ImageButton.setOnCopyListener(text: String) {
//                setOnClickListener { dataViewModel.copyText(text) }
//                setBackgroundColor(settings.layoutBackgroundColor)
//                imageTintList = ColorStateList.valueOf(settings.fontColor)
//            }
//
//            private fun TextViewFieldBinding.setText(text: String) {
//                textView.text = text
//            }
//
//
//            init {
//                when (moreInfoBinding) {
//                    is MoreWebsiteBinding -> {
//                        moreInfoBinding.buttonOpenUrl.setOnClickListener {
//                            openUrl()
//                        }
//
//                        moreInfoBinding.password.buttonVisibility.show()
//                        moreInfoBinding.password.buttonVisibility.setOnClickListener {
//                            updatePasswordView()
//                        }
//
//                        updatePasswordView(isPasswordVisible)
//                    }
//                    is MoreBankCardBinding -> {
//                        //todo доделать bankCard binding
//                    }
//                }
//            }
//
//            fun bindInfo(data: Data, position: Int) {
//                this.data = data
//                this.pos = position
//
//                when (moreInfoBinding) {
//                    is MoreWebsiteBinding -> {
//                        if (data !is Website) return
//
//                        moreInfoBinding.run {
//                            login.setText(data.login)
//                            password.setText(data.password)
//                            comment.setText(data.comment)
//
//                            if (data.nameAccount.isEmpty())
//                                nameAccount.hide()
//                            else
//                                nameAccount.txt = data.nameAccount
//
//                            if (data.comment.isEmpty()) {
//                                comment.textView.hide()
//                                comment.fieldItemButtonCopy.hide()
//                                commentHead.hide()
//                            }
//
//                            createBottomSheet()
//
//                            //colors
//                            root.backgroundTintList =
//                                ColorStateList.valueOf(settings.layoutBackgroundColor)
//
//                            for (item in arrayOf(login, password, comment)) {
//                                item.textView.setBackgroundColor(settings.layoutBackgroundColor)
//                                item.textView.setTextColor()
//                            }
//
//                            login.fieldItemButtonCopy.setOnCopyListener(data.login)
//                            password.fieldItemButtonCopy.setOnCopyListener(data.password)
//                            comment.fieldItemButtonCopy.setOnCopyListener(data.comment)
//
//                            loginHead.setTextColor()
//                            passwordHead.setTextColor()
//                            commentHead.setTextColor()
//
//                            nameAccount.setBackgroundColor(settings.layoutBackgroundColor)
//
//                            password.buttonVisibility.setBackgroundColor(settings.layoutBackgroundColor)
//
//                            password.buttonVisibility.imageTintList =
//                                ColorStateList.valueOf(settings.fontColor)
//
//                            buttonOpenUrl.setBackgroundResource(settings.buttonRes)
//
//                            root.setOnClickListener {
//                                if (settings.baseSettings.isUsingBottomView)
//                                    bottomDialogFragment.show(childFragmentManager)
//                            }
//                        }
//                    }
//                }
//            }
//
//
//            private fun updatePasswordView() {
//                isPasswordVisible = !isPasswordVisible
//                updatePasswordView(isPasswordVisible)
//            }
//
//            private fun updatePasswordView(visibility: Boolean) =
//                (moreInfoBinding as MoreWebsiteBinding).apply {
//                    if (visibility) {
//                        password.textView.inputType = InputType.TYPE_CLASS_TEXT
//                        password.buttonVisibility.setImageResource(R.drawable.visibility_off)
//                        password.buttonVisibility.contentDescription =
//                            getString(R.string.hide_password)
//                    } else {
//                        password.textView.inputType = 129
//                        password.buttonVisibility.setImageResource(R.drawable.visibility_on)
//                        password.buttonVisibility.contentDescription =
//                            getString(R.string.show_password)
//                    }
//                }
//
//            //todo нужно ли в bankCard?
//            fun createBottomSheet() {
//                moreInfoBinding as MoreWebsiteBinding
//
//                val heading = if (moreInfoBinding.nameAccount.isEmpty()) {
//                    (data as Website).login
//                }
//                else moreInfoBinding.nameAccount.text.toString()
//
//                bottomDialogFragment.setHeading(heading, beautifulDesign = true)
//
//                bottomDialogFragment.addView(R.drawable.edit, R.string.edit) {
//                    bottomDialogFragment.dismiss()
//                    startActivity(PasswordActivity
//                        .getIntent(context, (data as Website).address, pos))
//                }
//            }
//
//            fun openUrl() {
//                val address = (data as Website).address.createUrlString()
//
//                if (URLUtil.isValidUrl(address)) {
//                    val intent = Intent(Intent.ACTION_VIEW, address.toUri())
//                    startActivity(intent)
//                } else
//                    showToast(activity, "Неправильный адрес!")
//            }
//        }
//    }
}